x-pack/plugins/cases/public/common/hooks.ts

```diff
@@ -5,11 +5,97 @@
  * 2.0.
  */
 
+import { useCallback, useEffect, useState } from 'react';
+
+import { useLocation, useHistory } from 'react-router-dom';
+
+import useLocalStorage from 'react-use/lib/useLocalStorage';
+import { stringify } from 'query-string';
 import { STACK_APP_ID } from '../../common/constants';
 import { useCasesContext } from '../components/cases_context/use_cases_context';
 
+import { DEFAULT_QUERY_PARAMS } from '../containers/use_get_cases';
+import { parseUrlQueryParams } from './utils';
+
+import type { LocalStorageQueryParams, QueryParams, UrlQueryParams } from '../../common/ui/types';
+
 export const useIsMainApplication = () => {
   const { appId } = useCasesContext();
 
   return appId === STACK_APP_ID;
 };
+
+export function useUrlState(isSelectorView: boolean = false) {
+  const location = useLocation();
+  const history = useHistory();
+
+  const [queryParams, setQueryParams] = useState<UrlQueryParams>({});
+
+  const [localStorageQueryParams, setLocalStorageQueryParams] =
+    useLocalStorage<LocalStorageQueryParams>('cases.list.preferences');
+
+  const persistAndUpdateQueryParams = useCallback(
+    (params) => {
+      const urlParams: UrlQueryParams = parseUrlQueryParams(location.search);
+      const newQueryParams: QueryParams = { ...DEFAULT_QUERY_PARAMS };
+
+      newQueryParams.perPage =
+        params.perPage ??
+        urlParams.perPage ??
+        localStorageQueryParams?.perPage ??
+        DEFAULT_QUERY_PARAMS.perPage;
+      newQueryParams.sortField =
+        params.sortField ?? queryParams.sortField ?? DEFAULT_QUERY_PARAMS.sortField;
+      newQueryParams.sortOrder =
+        params.sortOrder ??
+        urlParams.sortOrder ??
+        localStorageQueryParams?.sortOrder ??
+        DEFAULT_QUERY_PARAMS.sortOrder;
+      newQueryParams.page = params.page ?? urlParams.page ?? DEFAULT_QUERY_PARAMS.page;
+
+      if (stringify(queryParams) !== stringify(newQueryParams)) {
+        const newLocalStorageQueryParams = {
+          perPage: newQueryParams.perPage,
+          sortOrder: newQueryParams.sortOrder,
+        };
+
+        setLocalStorageQueryParams(newLocalStorageQueryParams);
+
+        const newUrlParams = {
+          page: newQueryParams.page,
+          ...newLocalStorageQueryParams,
+        };
+
+        history.replace({
+          ...location,
+          search: stringify(newUrlParams),
+        });
+
+        setQueryParams(newQueryParams);
+      }
+    },
+    [history, localStorageQueryParams, location, setLocalStorageQueryParams, queryParams]
+  );
+
+  const setUrlQueryParams = useCallback(
+    (params) => {
+      if (isSelectorView) {
+        setQueryParams(params);
+      } else {
+        persistAndUpdateQueryParams(params);
+      }
+    },
+    [persistAndUpdateQueryParams, isSelectorView]
+  );
+
+  useEffect(() => {
+    setUrlQueryParams(isSelectorView ? DEFAULT_QUERY_PARAMS : {});
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Is this so we only render once? Do we run into issues if we have the dependencies here?

Reply:
Yup, this is so the URL is updated automatically the first time the page is opened.

I tried populating the array with all dependencies, which would then be triggered more than once. I wanted to avoid that.